File: golclinics/arrays-snd-strings/reverseSentence.js
const A = ["t", "h", "i", "s", " ", "i", "s", " ", "g", "o", "o", "d"];
// Intuition behind my solution :
// we start by reversing the whole sentence then reverse the indivisual words
function reverseSentence(A) {
  helperReverse(A, 0, A.length - 1);
  let startInd = 0;
  for (let i = 0; i <= A.length; i++) {
    if (A[i] === " " || i === A.length) {
      helperReverse(A, startInd, i - 1);
      startInd = i + 1;
    }
  }
  return A;
}

// helper function to implement reverse
function helperReverse(array, start, end) {
  while (start < end) {
    let temp = array[start];
    array[start] = array[end];
    array[end] = temp;
    start++;
    end--;
  }
}

console.log(reverseSentence(A));

File: Shizler/Arrays and Strings 1/Arrays/reverseSentence.py
def reversedSentence(A):
    def reverseElements(A,start=0,stop=len(A)-1):
        while start < stop:
            A[start], A[stop] = A[stop], A[start]
            start, stop = start+1, stop-1

    # reverse whole array
    reverseElements(A)
    start = 0

    # loop through whole array reversing subsets of it(words)
    while True:
        try:
            end = A.index(' ',start,len(A))
        except:
            reverseElements(A,start,len(A)-1)
            break

        reverseElements(A, start,end-1)

        start = end + 1

    print(A)


A = ['t','h','i','s',' ','i','s',' ','g','o','o','d']
reversedSentence(A)
